### PR TITLE
[WIP] Refactor HttpClient and MultipartContent with async/await

### PR DESCRIPTION
**Pull request by Maxwe11**

### CODE REVIEW

**Comment by Maxwe11** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@Maxwe11, thanks for taking a stab at this.  My initial reaction is that this doesn't go nearly far enough, as evidenced by the TaskCompletionSources that are still being passed around and that this hasn't really changed the structure of the code, but rather simply done localized changes to replace ContinueWiths in a method with awaits in the same method.  My expectation is that doing this robustly will result in a significant rewrite, and that very few (if any TaskCompletionSources) will be left at the end across the codebase.

**Comment by Maxwe11** [comment]
@stephentoub thanks for the initial feedback, I've updated PR. Now only a couple TaskCompletionSources left.

File: src/System.Net.Http/src/System/Net/Http/HttpClient.cs [hunk 1]
```diff
@@ -157,44 +157,128 @@ public Task<Stream> GetStreamAsync(string requestUri)
...
                 }
-                HttpResponseMessage response = requestTask.Result;
-                if (response.Content == null)
+                else
                 {
-                    tcs.TrySetResult(defaultValue);
-                    return;
+                    // Otherwise option is HttpCompletionOption.ResponseHeadersRead.
+                    // This option used by GetStreamAsync so we expecting that generic argument is Stream.
+                    Debug.Assert(typeof(T) == typeof(Stream), "We expection Stream as generic argument at this branch");
```
Comment (Maxwe11) [inline src/System.Net.Http/src/System/Net/Http/HttpClient.cs]: Will fix typo

**Comment by Maxwe11** [comment]
@dotnet-bot test this please

**Pull request closed without merging**